### ISSUE TITLE
how do I find if the location(GPS) enable or disable on background mode?

### ISSUE THREAD

**Issue by bhanuka96** [transistorsoft/flutter_background_geolocation#503]
## Your Environment
* Plugin version: 1.11.0
* Platform: iOS or Android
* OS version: any OS
* Device manufacturer / model: any Model
* Flutter info (`flutter doctor`):
* Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.6, on macOS 11.1 20C69 darwin-x64, locale en-LK)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.52.1)
[✓] Connected device (1 available)

* Plugin config:
```dart <-- Syntax highlighting: DO NOT REMOVE -->
bg.Config(
        desiredAccuracy: bg.Config.DESIRED_ACCURACY_NAVIGATION,
        distanceFilter: 0.0,
        locationUpdateInterval: 10000,
        locationTimeout: 60,
        persistMode: 1,
        desiredOdometerAccuracy: 10,
        stationaryRadius: 25,
        foregroundService: true,
        disableMotionActivityUpdates: true,
        enableHeadless: true,
        stopOnTerminate: false,
        startOnBoot: true,
        disableStopDetection: true,
        pausesLocationUpdatesAutomatically: false,
        allowIdenticalLocations: true,
        useSignificantChangesOnly: false,
        transistorAuthorizationToken: token,
        debug: kDebugMode,
        logLevel: kReleaseMode ? bg.Config.LOG_LEVEL_OFF : bg.Config.LOG_LEVEL_VERBOSE,
        notificationPriority: kReleaseMode ? -2 : 1,
        showsBackgroundLocationIndicator: kDebugMode,
        locationAuthorizationRequest: 'Always',
        backgroundPermissionRationale: PermissionRationale();
```

I am currently using this way to capture location status.

```
    bg.BackgroundGeolocation.onProviderChange((bg.ProviderChangeEvent event) async {
      if (event?.gps != null) {
        _gpsController.add(event);
      }
    });
```

Is it possible to identify background mode too? I am using androidheadles too.

```
void headlessTask(bg.HeadlessEvent headlessEvent) async {
  switch (headlessEvent.name) {
    case bg.Event.LOCATION:
      {
       //logic
        break;
      }
    case bg.Event.TERMINATE:
      bg.State state = headlessEvent.event;
      print('- State: $state');
      break;
    case bg.Event.ENABLEDCHANGE:
      bool enabled = headlessEvent.event;
      print('EnabledChangeEvent: $enabled');
      break;
  }
}
```

**Comment by christocracy** [comment]
The Flutter framework already provides a mechanism for determining when an app is in the background with [LifeCycleObserver](https://medium.com/@marcoleong.developer/lifecycle-in-flutter-widgetsbindingobserver-2eb782236f8b)

When your app is running *Headless*, your app *necessarily* in the background.

**Comment by bhanuka96** [comment]
@christocracy Thanks. I mean, when the app running in the background (app closes from the recent menu).

**Comment by christocracy** [comment]
> I mean, when the app running in the background (app closes from the recent app menu in device)

There are 3 possible states for an app.
- *foreground*
- *background*
- *terminated*.

When the Android headless-task is receiving events, your app is *necessarily* terminated and operating in the background.

**Comment by bhanuka96** [comment]
@christocracy I mean, terminated.

**Comment by bhanuka96** [comment]
@christocracy Is it possible to find location(GPS) enable or disable on terminated mode?

**Comment by christocracy** [comment]
When the device is moving and recording locations, those locations will be sent to your registered headless-task.

If you want to disable tracking, simply call `.stop()`.

If you want to periodically call `getCurrentPosition`, use `react-native-background-fetch` (already included as a dependency).

**Comment by bhanuka96** [comment]
@christocracy Extremely sorry for my explanation. I mean this location(GPS). https://drive.google.com/file/d/1mIVC97DeLTj4oTnM5brcFFkBmXXbktCl/view?usp=sharing
please skip the first 10 seconds.

I used the below way to capture this on the app active mode, but I cannot capture it when the app runs on terminated mode. Is it possible to capture it?

```
 bg.BackgroundGeolocation.onProviderChange((bg.ProviderChangeEvent event) async {
      if (event?.gps != null) {
        _gpsController.add(event);
      }
    });
```

**Comment by christocracy** [comment]
`providerchange` events are sent to your registered headless-task while the Android app is *terminated*.

Are you observing `$ adb logcat *:S TSLocationManager`?  To use the plugin, you *must* observe the logs at all times.

**Comment by bhanuka96** [comment]
@christocracy Thank you. it's not mentioned on your wiki. https://github.com/transistorsoft/flutter_background_geolocation/wiki/Android-Headless-Mode

**Comment by christocracy** [comment]
Why not simply `print(event)` and observe what events arrive?

**Comment by bhanuka96** [comment]
@christocracy I just want to tell you about this. That's it.

**Comment by bhanuka96** [comment]
@christocracy 

I added `bg.Event.PROVIDERCHANGE:` but it's not printed. and I commented code inside `bg.Event.PROVIDERCHANGE:`. How is `provider` printed inside location? 

```
void headlessTask(bg.HeadlessEvent headlessEvent) async {
    print('event : ${headlessEvent.name}');
    switch (headlessEvent.name) {
      case bg.Event.LOCATION:
        {
          try {
            locationAdded(headlessEvent);
          } catch (e) {
            print('Error headless : $e');
          }
          break;
        }

      // case bg.Event.PROVIDERCHANGE:
      //   {
      //     bg.ProviderChangeEvent event = headlessEvent.event;
      //     print('- Provider: $event');
    }
}

void locationEvent(bg.HeadlessEvent headlessEvent) async {
  bg.Location location = headlessEvent.event;
  print('- Location: $location');
  if(location.toMap().containsKey('provider')){
    print('- Location Provider: ${location.toMap()['provider']}');
  }
}
```


These are my log messages

**Switch on location in device**

`- Location: [Location {is_moving: true, uuid: uuid, timestamp: 2021-01-29T17:32:34.000Z, odometer: 0, coords: {latitude: 0, longitude: 0, accuracy: 27.9, speed: 0.18, speed_accuracy: 0.22, heading: 20.05, heading_accuracy: -1, altitude: -58, altitude_accuracy: 24}, activity: {type: still, confidence: 100}, battery: {is_charging: true, level: 1}, extras: {}}]`

**Now, I am turn off location in my device**

`- Location: [Location {event: providerchange, provider: {network: true, gps: false, enabled: true, status: 3, accuracyAuthorization: 0}, is_moving: true, uuid: uuid, timestamp: 2021-01-29T17:35:30.113Z, odometer: 0, coords: {latitude: 0, longitude: 0, accuracy: 56.1, speed: -1, speed_accuracy: -1, heading: -1, heading_accuracy: -1, altitude: -1, altitude_accuracy: -1}, activity: {type: still, confidence: 100}, battery: {is_charging: true, level: 1}, extras: {}}]`

`I/flutter ( 9114): - Location Provider: {network: true, gps: false, enabled: true, status: 3, accuracyAuthorization: 0}`
